azext_iot/deviceupdate/providers/base.py

```diff
@@ -0,0 +1,190 @@
+                    )
+                    has_system = True
+                else:
+                    user_assigned_identities[identity] = {}
+
+            return DeviceUpdateMgmtModels.ManagedServiceIdentity(
+                type=target_identity_type,
+                user_assigned_identities=user_assigned_identities,
+            )
+
+    def assign_msi_scope(
+        self,
+        principal_id: str,
+        scope: str,
+        principal_type: str = "ServicePrincipal",
+        role: str = "Contributor",
+    ) -> dict:
+        assign_op = self.cli.invoke(
+            f"role assignment create --scope '{scope}' --role '{role}' --assignee-object-id '{principal_id}' "
+            f"--assignee-principal-type '{principal_type}'"
+        )
+        if not assign_op.success():
+            raise CLIInternalError(f"Failed to assign '{principal_id}' the role of '{role}' against scope '{scope}'.")
+
+        return assign_op.as_json()
+
+    def get_rg_location(
+        self,
+        resource_group_name: str,
+    ):
+        resource_group_meta = self.cli.invoke(
+            f"group show --name {resource_group_name}"
+        ).as_json()
+        return resource_group_meta["location"]
+
+
+class DeviceUpdateInstanceManager(DeviceUpdateAccountManager):
+    def __init__(self, cmd):
+        super().__init__(cmd=cmd)
+
+    def assemble_iothub_resources(
+        self, resource_ids: List[str]
+    ) -> List[DeviceUpdateMgmtModels.IotHubSettings]:
+        iothub_settings_list: List[DeviceUpdateMgmtModels.IotHubSettings] = []
+        for id in resource_ids:
+            iothub_settings_list.append(
+                DeviceUpdateMgmtModels.IotHubSettings(resource_id=id)
+            )
+        return iothub_settings_list
+
+    def assemble_diagnostic_storage(
+        self, storage_id: str
+    ) -> DeviceUpdateMgmtModels.DiagnosticStorageProperties:
+        diagnostic_storage = DeviceUpdateMgmtModels.DiagnosticStorageProperties(
+            authentication_type="KeyBased", resource_id=storage_id
+        )
+        cstring_op = self.cli.invoke(
+            f"storage account show-connection-string --ids {storage_id}"
+        )
+        if not cstring_op.success():
+            raise CLIInternalError(f"Failed to fetch storage account connection string with resource id of '{storage_id}'.")
+        diagnostic_storage.connection_string = cstring_op.as_json()["connectionString"]
+
+        # @digimaun - the service appears to have a limitation handling the EndpointSuffix segment, it must be at the end.
+        split_cstring: list = diagnostic_storage.connection_string.split(";")
+        endpoint_suffix = "EndpointSuffix=core.windows.net"
+        for i in range(0, len(split_cstring)):
+            if "EndpointSuffix=" in split_cstring[i]:
+                endpoint_suffix = split_cstring.pop(i)
+                break
+        diagnostic_storage.connection_string = (
+            f'{";".join(split_cstring)};{endpoint_suffix}'
```

Comment:
why not just add the endpoint suffix to the end of the split cstring and then use one join?